test/e2e/rule_test.go

```diff
@@ -436,6 +447,89 @@ func TestRule(t *testing.T) {
 	})
 }
 
+// TestRule_CanRemoteWriteData checks that Thanos Ruler can be run in stateless mode
+// where it remote_writes rule evaluations to a Prometheus remote-write endpoint (typically
+// a Thanos Receiver).
+func TestRule_CanRemoteWriteData(t *testing.T) {
+	t.Parallel()
+
+	e, err := e2e.NewDockerEnvironment("e2e_test_rule_remote_write")
+	testutil.Ok(t, err)
+	t.Cleanup(e2ethanos.CleanScenario(t, e))
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
+	t.Cleanup(cancel)
+
+	rulesSubDir := "rules"
+	rulesPath := filepath.Join(e.SharedDir(), rulesSubDir)
+	testutil.Ok(t, os.MkdirAll(rulesPath, os.ModePerm))
+
+	for i, rule := range []string{testRuleRecordAbsentMetric, testAlertRuleWarnOnPartialResponse} {
+		createRuleFile(t, filepath.Join(rulesPath, fmt.Sprintf("rules-%d.yaml", i)), rule)
+	}
+
+	am, err := e2ethanos.NewAlertmanager(e, "1")
+	testutil.Ok(t, err)
+	testutil.Ok(t, e2e.StartAndWaitReady(am))
+
+	receiver, err := e2ethanos.NewIngestingReceiver(e, "1")
+	testutil.Ok(t, err)
+	testutil.Ok(t, e2e.StartAndWaitReady(receiver))
+	rwURL := mustURLParse(t, e2ethanos.RemoteWriteEndpoint(receiver.InternalEndpoint("remote-write")))
+
+	q, err := e2ethanos.NewQuerierBuilder(e, "1", receiver.InternalEndpoint("grpc")).Build()
+	testutil.Ok(t, err)
+	testutil.Ok(t, e2e.StartAndWaitReady(q))
+	r, err := e2ethanos.NewStatelessRuler(e, "1", rulesSubDir, []alert.AlertmanagerConfig{
+		{
+			EndpointsConfig: httpconfig.EndpointsConfig{
+				StaticAddresses: []string{
+					am.InternalEndpoint("http"),
+				},
+				Scheme: "http",
+			},
+			Timeout:    amTimeout,
+			APIVersion: alert.APIv1,
+		},
+	}, []httpconfig.Config{
+		{
+			EndpointsConfig: httpconfig.EndpointsConfig{
+				StaticAddresses: []string{
+					q.InternalEndpoint("http"),
+				},
+				Scheme: "http",
+			},
+		},
+	}, &config.RemoteWriteConfig{
+		URL:  &common_cfg.URL{URL: rwURL},
+		Name: "thanos-receiver",
+	})
+	testutil.Ok(t, err)
+	testutil.Ok(t, e2e.StartAndWaitReady(r))
+
+	t.Run("can fetch remote-written samples from receiver", func(t *testing.T) {
+		testRecordedSamples := "test_absent_metric"
+		queryAndAssertSeries(t, ctx, q.Endpoint("http"), testRecordedSamples, func() time.Time {
+			return time.Now()
+		}, promclient.QueryOptions{
+			Deduplicate: false,
+		}, []model.Metric{
+			{
+				"__name__":  "test_absent_metric",
+				"job":       "thanos-receive",
+				"receive":   "1",
+				"tenant_id": "default-tenant",
+			},
+		})
+	})
+}
+
+// TestRule_CanPersistWALData checks that in stateless mode, Thanos Ruler can persist rule evaluations
+// which couldn't be sent to the remote write endpoint (e.g because receiver isn't available).
+func TestRule_CanPersistWALData(t *testing.T) {
+	//TODO: Implement test with unavailable remote-write endpoint(receiver)
```

Comment:
Do we need to implement this before merging?

Reply:
I guess this is not a blocker as we are using the upstream prom agent code already so data should be persisted. Not sure how to implement this right now though.

Reply:
:+1: